### PR TITLE
NEW FormField Title tooltip

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-asset-admin/issues/1214

### CODE REVIEW

File: client/src/styles/legacy/_forms.scss
Comment (maxime-rainville) [inline client/src/styles/legacy/_forms.scss:154]: This should be in a SCSS file in the same folder as Tip.

**Review by maxime-rainville — CHANGES_REQUESTED** [review]
Need to clarify if Label tip and Field tip can be use simultaneously. If not, TextField should suppress the label tip somehow.

My instinct would be to convert Tip to be a functional components, but that's not a hill I'm desperate to die on.

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: Updated

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: Updated both the tip types to use the Silverstripe Button component

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: Now they're both using the Silverstripe Button component, the only difference between them is the classnames they use, so I've kept them in the same file

File: client/src/components/FieldHolder/FieldHolder.js
Comment (emteknetnz) [reply client/src/components/FieldHolder/FieldHolder.js]: Updated

File: client/src/components/FieldHolder/FieldHolder.js
Comment (emteknetnz) [reply client/src/components/FieldHolder/FieldHolder.js]: Updated

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: Updated

File: client/src/styles/legacy/_forms.scss
Comment (emteknetnz) [reply client/src/styles/legacy/_forms.scss]: Moved to its own scss file

**Review by maxime-rainville — COMMENTED** [review]
@clarkepaul Seems like maybe we should only have one of those tool tip open at a time. Not sure what would happen if you ended up in a situation where one tool tip overlap with another. Also, there's no mouse or focus styles for the `(?)` icon.

image

File: client/src/components/Tip/Tip.scss
Comment (maxime-rainville) [inline client/src/components/Tip/Tip.scss:15]: You could probably avoid this by conditionally changing the button `color` prop to `link`.

File: client/src/components/Tip/tests/Tip-story.js
Comment (maxime-rainville) [inline client/src/components/Tip/tests/Tip-story.js:50]: Given how this story is structured, having this prop here doesn't provide any value. 

We should have a tip on the label and one on the field. Maybe we need a different story for each use case. Or maybe this should be on a story specific to form fields.

**Review by maxime-rainville — CHANGES_REQUESTED** [review]

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js:12]: The pre-existing name is "importance level", so I think "type" isn't confusing and makes the most sense

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js:13]: I think "title" makes more sense than say label because the the tip is next to the form field title, rather than the `<label>` html tag, which to me seems less obvious since it requires more knowledge of the implementation

If we change the name here we'd want to be consistent all the way back to `FormField::setLabelTip()`, which doesn't seems less obvious to me what it does

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: Updated

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: I've refactored quite a lot of Tip.js

File: client/src/components/Tip/Tip.scss
Comment (emteknetnz) [reply client/src/components/Tip/Tip.scss:23]: Changing the color prop on the `<Button>` component doesn't fix this

File: client/src/components/Tip/tests/Tip-story.js
Comment (emteknetnz) [reply client/src/components/Tip/tests/Tip-story.js]: I've split the two different types to be different things in the pattern-lib

**Comment by emteknetnz** [comment]
@clarkepaul I've refactored the Tip so that it uses the 'legacy' Popover trigger - see https://reactstrap.github.io/components/popovers/ - what this means is that when you click outside of a popover it will close the old one.  This means that only one popover at a time can be open.

I've added focus/actives states where I've changed the icon color because it's fairly easy to do.  Borders and box shadows are difficult since it's a round icon - the border/box-shadow will display as a square

I've added focus state when tabbing with a keyboard and mouse hover. 

It uses the `$input-focus-border-color` which is used for the text field border color when focusing in a field

image

I've used `darken($input-focus-border-color, 15%);` for the active state when clicking the icon

image

Let me know if you'd like these colors changed

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: Updated

**Comment by clarkepaul** [comment]
@emteknetnz @maxime-rainville nice that you got the popovers to close, that was put into the too hard basket a while back by @Cheddam so I'd think that there is an issue floating around for it.

@emteknetnz I'll get back to you on the colours of focus, active soon.

**Comment by clarkepaul** [comment]
@emteknetnz I can't think of any other patterns we've used which would be any good apart from colour so I think you've gone in the right direction. Can you use the link colour (should output at #0071c4) for focus and active, use link:hover colour (should output at #0062ab) for just when it is hovered, they should contrast better.

**Comment by emteknetnz** [comment]
@clarkepaul 
image
image
image

**Review by maxime-rainville — CHANGES_REQUESTED** [review]
@clarkepaul Here's a video that shows the tooltip after and before: https://youtu.be/XbfIa7kdN1s 

One thing to note is that the tool tip will now be dismissed once you click somewhere else on the screen. I'm calling this an improvement over the status quo.

Tested in IE 11 and it worked great there as well.

@emteknetnz Massive improvement! I'm still not happy with having two things reffered as `type` even if it doesn't affect the API we expose to the rest of the system. Rename one of those things. But otherwise it looks good to merge pending Paul's approval.

File: client/src/components/Tip/Tip.js
Comment (maxime-rainville) [inline client/src/components/Tip/Tip.js:62]: What's the point of onClick here?

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js:12]: I've changed the tip importance level from 'type' to 'description'

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js]: Updated

File: client/src/components/Tip/Tip.js
Comment (emteknetnz) [reply client/src/components/Tip/Tip.js:62]: The button component (presumably the reactstrap one) wants an onClick function defined otherwise it doesn't want to work.  So I put a noop in

**Review by maxime-rainville — CHANGES_REQUESTED** [review]
Just two last tweaks and we should be good

**Comment by emteknetnz** [comment]
@maxime-rainville done

File: client/src/components/Tip/Tip.js
Comment (maxime-rainville) [reply client/src/components/Tip/Tip.js:50]: @emteknetnz Doesn't look like this one was done. Translation won't work if it's not actioned.

**Review by maxime-rainville — APPROVED** [review]
Looks beautiful!

**Pull request merged**
